### PR TITLE
UI rework

### PR DESCRIPTION
**Pull request by headassbtw**
Reworks the UI bridge code in a lot of ways:

- Blitzes the games in, no waiting on image/detail loading
- FIXED THE GAME IMAGE SYSTEM (now called `UIImage`, subject to change) I hated it so much and now it actually makes sense to use
- Game images and details now async load when the game is selected, and show a placeholder for a skeleton when they're not loaded
- Games now show the system requirements
- Probably some other cleanup things I forgot, I feel like I'm actually in control of my codebase now
- Fixed the app BG vignette on high-DPI configurations

### CODE REVIEW

File: maxima-ui/src/util/markdown.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            + &segment[beglink..endlink].to_string()
+            + ")";
+        rtn = rtn[..idx + (beglink - 9)].to_string()
+            + &modlink
+            + &rtn[idx + endtext + 4..].to_string();
+
+        idx += (beglink - 9) + (modlink.len());
+    }
+    idx = 0; // image removal, easymark doesn't support them and EAD doesn't seem to display them
+    while rtn.len() > idx && rtn[idx..].find("<img").is_some() && rtn[idx..].find(">").is_some() {
```
Comment (battledash) [inline maxima-ui/src/util/markdown.rs]: Add newlines to separate the idx line here

File: maxima-ui/src/util/markdown.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        .replace("*", "\\*")
+        .replace("<br>", "\n")
+        .replace("<b>", "*")
+        .replace("</b>", "*")
+        .replace("<span style=\"font-size: small;\">", "") // WE DO NOT CARE 🗣️
+        .replace("</span>", "")
+        .replace("&nbsp;", " ")
+        .replace("\\", "");
+    let mut idx: usize = 0;
+    // link corrector
```
Comment (battledash) [inline maxima-ui/src/util/markdown.rs]: Add a newline before the comment here

File: maxima-ui/src/util/markdown.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+        let begimg = segment.find("<img").expect("welp") + 4;
+        let endimg = segment[begimg..].find(">").expect("welp") + begimg;
+
+        rtn = rtn[..idx + (begimg - 4)].to_string() + &rtn[idx + endimg + 1..].to_string();
+
+        idx += begimg - 4;
+    }
+
+    rtn.replace("/", "\\/") // links are caught in the crossfire of this, but clicking them still leads to where they shouold
```
Comment (battledash) [inline maxima-ui/src/util/markdown.rs]: typo in the comment

File: maxima-ui/src/views/game_view.rs [hunk 4]
```diff
@@ -26,16 +27,41 @@ pub struct GameViewBar {
...
+      game.images = GameUIImagesWrapper::Loading;
+      None
+    },
+    GameUIImagesWrapper::Loading => {
+      None
+    },
+    GameUIImagesWrapper::Available(images) => {
+      Some(images) },
+  };
+  let game_details: Option<&GameDetails> = match &game.details {
```
Comment (battledash) [inline maxima-ui/src/views/game_view.rs]: Newline here

File: maxima-ui/src/views/game_view.rs [hunk 5]
```diff
@@ -106,11 +132,15 @@ pub fn game_view_details_panel(app : &mut DemoEguiApp, ui: &mut Ui) {
...
-                        stats.label(RichText::new(&app.locale.localization.games_view.main.playtime).color(Color32::BLACK).strong());
-                        stats.label(RichText::new(format!(": {:?} hours",app.games[app.game_sel].time as f32 / 10.0)).color(Color32::BLACK));
-                        stats.separator();
+                        if let Some(details) = game_details {
+                          stats.label(RichText::new(&app.locale.localization.games_view.main.playtime).color(Color32::BLACK).strong());
+                          stats.label(RichText::new(format!(": {:?} hours",details.time as f32 / 10.0)).color(Color32::BLACK));
+                          stats.separator();
                         stats.label(RichText::new(&app.locale.localization.games_view.main.achievements).color(Color32::BLACK).strong());
-                        stats.label(RichText::new(format!(": {:?} / {:?}",app.games[app.game_sel].achievements_unlocked,app.games[app.game_sel].achievements_total)).color(Color32::BLACK));
+                        stats.label(RichText::new(format!(": {:?} / {:?}",details.achievements_unlocked,details.achievements_total)).color(Color32::BLACK));
```
Comment (battledash) [inline maxima-ui/src/views/game_view.rs]: These are some very long lines

File: maxima-ui/src/views/game_view.rs [hunk 6]
```diff
@@ -200,40 +230,57 @@ pub fn game_view_details_panel(app : &mut DemoEguiApp, ui: &mut Ui) {
...
-              ui.put(logo_rect, egui::Image::new(logo, logo_size));
-            } else {
-              ui.put(logo_rect, egui::Spinner::new().size(logo_size.min_elem()/2.0));
-              //ui.add_sized(logo_rect.size(), egui::Spinner::new());
-              //ui.painter().rect_filled(logo_rect, Rounding::same(0.0), Color32::TRANSPARENT);
-            }
+          if let Some(images) = game_images {
+            if let Some(logo) = &images.logo {
+              let logo_size_pre = 
+              if logo.size.x >= logo.size.y {
```
Comment (battledash) [inline maxima-ui/src/views/game_view.rs]: This shouldn't be 2 lines

File: maxima-ui/src/main.rs [hunk 7]
```diff
@@ -739,6 +741,18 @@ impl eframe::App for DemoEguiApp {
...
+                    let mut warning_margin = Margin::same(0.0 - APP_MARGIN.x);
+                    warning_margin.bottom = APP_MARGIN.y;
+                    egui::Frame::default()
+                    .fill(Color32::RED)
+                    .outer_margin(warning_margin)
+                    .show(ui, |ui| {
+                        ui.vertical_centered(|ui| {
+                            ui.heading(egui::RichText::new(&self.locale.localization.errors.critical_thread_crashed).color(Color32::BLACK).size(16.0));
+                        });
+                    });
```
Comment (battledash) [inline maxima-ui/src/main.rs]: This should be indented, and the huge line should be split up

File: maxima-ui/src/translation_manager.rs [hunk 8]
```diff
@@ -84,6 +85,12 @@ pub struct LocalizedGamesViewMain {
     pub no_loaded_games : String
 }
 
+#[derive(Deserialize)]
+pub struct LocalizedGamesViewDetails {
+    pub min_system_req : String,
+    pub rec_system_req : String,
```
Comment (battledash) [inline maxima-ui/src/translation_manager.rs]: There shouldn't be a space before the colons here

**Review by battledash — CHANGES_REQUESTED** [review]

File: maxima-ui/src/bridge/login_creds.rs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        let user = maxima.local_user().await?;
+        let lmessage = MaximaLibResponse::LoginResponse(InteractThreadLoginResponse {
+            success: true,
+            description: user.player().as_ref().unwrap().display_name().to_owned(),
+        });
+        info!("Successfully logged in with username/password");
+        channel.send(lmessage)?;
+    }
```
Comment (battledash) [inline maxima-ui/src/bridge/login_creds.rs]: Why's this in a block?

File: maxima-ui/src/game_info_image_handler.rs [hunk 10]
```diff
@@ -1,206 +1,67 @@
...
+      error!("Failed to create directory {:?}", &cache_folder);
+    }
+  }
+  if !fs::metadata(&file_name).is_ok() { //image hasn't been cached yet
+    if let Some(img_url) = url {
+      info!("Downloading image at {:?}", img_url);
+      let result = reqwest::get(&img_url).await;
+      if let Ok(response) = result {
+        if let Ok(body) = response.bytes().await {
+          if let Ok(mut file) = File::create(&file_name).await {
```
Comment (battledash) [inline maxima-ui/src/game_info_image_handler.rs]: Please try to code guard this

File: maxima-ui/src/game_info_image_handler.rs [hunk 11]
```diff
@@ -1,206 +1,67 @@
...
-      } else {
-        handler.get_image(self.slug.clone(),
-        GameImageType::Logo,
-        Some(self.path.clone()),
-        Some(logo.url.clone()))
-      }
-    } else {
-      Err(Error::msg("Game does not have a logo"))
-    }
+  let fs_load = ImageLoader::load_from_fs(&file_name.to_str().unwrap());
```
Comment (battledash) [inline maxima-ui/src/game_info_image_handler.rs]: Newline here

File: maxima-ui/src/interact_thread.rs [hunk 12]
```diff
@@ -68,18 +73,26 @@ impl MaximaThread {
         let (tx1, rx0) = std::sync::mpsc::channel();
         let context = ctx.clone();
         tokio::task::spawn(async move {
+            let die_fallback_transmittter = tx1.clone();
+            //panic::set_hook(Box::new( |_| {}));
             let result = MaximaThread::run(rx1, tx1, &context).await;
             if result.is_err() {
+                let _ = die_fallback_transmittter
```
Comment (battledash) [inline maxima-ui/src/interact_thread.rs]: No need to make this a variable

File: maxima-ui/src/main.rs [hunk 13]
```diff
@@ -536,54 +548,6 @@ fn tab_button(ui: &mut Ui, edit_var: &mut PageType, page: PageType, label: Strin
...
-                            self.games[idx].logo = Some(assign.into());
-                        }
-                    }
-                }
-            }
-        } else {
-            //info!("lol, lmao");
-        }
-
         if let Ok(rcv) = self.backend.rx.try_recv() {
```
Comment (battledash) [inline maxima-ui/src/main.rs]: Can this be a separate function?

**Review by battledash — CHANGES_REQUESTED** [review]

File: maxima-ui/src/bridge/game_images.rs [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        .unwrap()
+        .join("cache/ui/images/")
+        .join(&slug)
+        .join("logo.png");
+    let has_hero = fs::metadata(&game_hero).is_ok();
+    let has_logo = fs::metadata(&game_logo).is_ok();
+    let images: Option<ServiceGame> = // TODO: make it a result
+                        if !has_hero || !has_logo { //game hasn't been cached yet
+                            let maxima = maxima_arc.lock().await;
+                            maxima.service_layer().request(SERVICE_REQUEST_GAMEIMAGES, ServiceGameImagesRequestBuilder::default().should_fetch_context_image(!has_logo).should_fetch_backdrop_images(!has_hero).game_slug(slug.clone()).locale(maxima.locale().short_str().to_owned()).build()?).await?
```
Comment (battledash) [inline maxima-ui/src/bridge/game_images.rs]: Make this request multiple lines, one line for each function call

File: maxima-ui/src/bridge/game_images.rs [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+        .join("hero.jpg");
+    let game_logo = maxima_dir()
+        .unwrap()
+        .join("cache/ui/images/")
+        .join(&slug)
+        .join("logo.png");
+    let has_hero = fs::metadata(&game_hero).is_ok();
+    let has_logo = fs::metadata(&game_logo).is_ok();
+    let images: Option<ServiceGame> = // TODO: make it a result
+                        if !has_hero || !has_logo { //game hasn't been cached yet
```
Comment (battledash) [inline maxima-ui/src/bridge/game_images.rs]: What's with the crazy indenting here

File: maxima-ui/src/bridge/game_images.rs [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+                            let maxima = maxima_arc.lock().await;
+                            maxima.service_layer().request(SERVICE_REQUEST_GAMEIMAGES, ServiceGameImagesRequestBuilder::default().should_fetch_context_image(!has_logo).should_fetch_backdrop_images(!has_hero).game_slug(slug.clone()).locale(maxima.locale().short_str().to_owned()).build()?).await?
+                        } else { None };
+    let hero_url: Option<String> = if has_hero {
+        debug!("Using cached hero image for {:?}", slug);
+        None
+    } else {
+        if let Some(imgs) = &images {
+            if let Some(pack) = &imgs.key_art() {
+                if let Some(img) = &pack.aspect_2x1_image() {
```
Comment (battledash) [inline maxima-ui/src/bridge/game_images.rs]: Are code guards possible here too?

File: maxima-ui/src/bridge/game_images.rs [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+            None
+        }
+    };
+    let logo_url: Option<String> = if has_logo {
+        debug!("Using cached logo for {:?}", slug);
+        None
+    } else {
+        if let Some(imgs) = &images {
+            if let Some(logos) = &imgs.primary_logo() {
+                if let Some(img) = &logos.largest_image() {
```
Comment (battledash) [inline maxima-ui/src/bridge/game_images.rs]: And here

File: maxima-ui/src/bridge/game_images.rs [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+                    None
+                }
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    };
+    let ctx = ctx.clone();
```
Comment (battledash) [inline maxima-ui/src/bridge/game_images.rs]: Newline here

File: maxima-ui/src/bridge/game_images.rs [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    let is_logo = logo_url.is_some() || has_logo;
+    tokio::task::spawn(async move {
+        // this looks weird but i was going to flip them anyway, heroes are always bigger, and putting the await inside of the optional logo makes sense
+        let hero = UIImage::load(
+            slug.clone(),
+            GameImageType::Hero,
+            if has_hero { None } else { hero_url },
+            ctx.clone(),
+        );
+        let logo = UIImage::load(
```
Comment (battledash) [inline maxima-ui/src/bridge/game_images.rs]: Newline here

File: maxima-ui/src/bridge/game_images.rs [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+            slug.clone(),
+            GameImageType::Logo,
+            if has_logo { None } else { logo_url },
+            ctx.clone(),
+        );
+        let hero = hero.await;
```
Comment (battledash) [inline maxima-ui/src/bridge/game_images.rs]: And here

File: maxima-ui/src/bridge/get_games.rs [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+    debug!("recieved request to load games");
+    let maxima = maxima_arc.lock().await;
+    let logged_in = maxima.auth_storage().lock().await.current().is_some();
+    if !logged_in {
+        info!("Ignoring request to load games, not logged in.");
+        return Err(Error::msg("Not logged in."));
+    }
+
+    let owned_games = maxima.owned_games(1).await.unwrap();
+    if let Some(games) = owned_games.owned_game_products() {
```
Comment (battledash) [inline maxima-ui/src/bridge/get_games.rs]:
```rust
let owned_game_products = owned_games.owned_game_products();
if owned_game_products.is_none() {
    return Ok(());
}

for game in owned_game_products.unwrap().items() {
```

File: maxima-ui/src/bridge/get_games.rs [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+    maxima_arc: Arc<Mutex<Maxima>>,
+    channel: Sender<MaximaLibResponse>,
+    ctx: &Context,
+) -> Result<()> {
+    debug!("recieved request to load games");
+    let maxima = maxima_arc.lock().await;
+    let logged_in = maxima.auth_storage().lock().await.current().is_some();
+    if !logged_in {
+        info!("Ignoring request to load games, not logged in.");
+        return Err(Error::msg("Not logged in."));
```
Comment (battledash) [inline maxima-ui/src/bridge/get_games.rs]: `bail!` here

File: maxima-ui/src/frontend_processor.rs [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+                    info!("Logged in as {}!", &res.description);
+                    app.user_name = res.description.clone();
+                    app.backend
+                        .tx
+                        .send(interact_thread::MaximaLibRequest::GetGamesRequest)
+                        .unwrap();
+                } else {
+                    warn!("Login failed.");
+                    app.in_progress_credential_status = res.description
+                }
```
Comment (battledash) [inline maxima-ui/src/frontend_processor.rs]:
```rust
if !res.success {
    warn!("Login failed.");
    app.in_progress_credential_status = res.description;
    continue;
}

...
```

File: maxima-ui/src/frontend_processor.rs [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+            }
+            interact_thread::MaximaLibResponse::GameInfoResponse(res) => {
+                app.games.push(res.game);
+                ctx.request_repaint(); // Run this loop once more, just to see if any games got lost
+            }
+            interact_thread::MaximaLibResponse::GameDetailsResponse(res) => {
+                for game in &mut app.games {
+                    if let Ok(resp) = &res.response {
```
Comment (battledash) [inline maxima-ui/src/frontend_processor.rs]:
```rust
if res.response.is_err() {
    continue;
}
```

File: maxima-ui/src/frontend_processor.rs [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+                        if game.slug == res.slug {
```
Comment (battledash) [inline maxima-ui/src/frontend_processor.rs]:
```rust
if game.slug != res.slug {
   continue;
}
```

File: maxima-ui/src/fs/image_loader.rs [hunk 26]
```diff
@@ -1,57 +1,73 @@
...
-        // but here we are. if it works, it works, and i sure as hell don't want to fix it.
-        let mut f = std::fs::File::open(path)?;
-        let mut buffer = String::new();
-        f.read_to_string(&mut buffer)?;
-        if let Ok(yeah) = RetainedImage::from_svg_str(format!("{:?}_Retained_Decoded",path), &buffer) {
-          return Ok(yeah)
+    pub fn load_from_fs(path: &str) -> Result<egui_extras::RetainedImage> {
+        debug!("Loading image {:?}", path);
+        if let Ok(img) = ImageReader::open(path) {
+            if let Ok(img_decoded) = img.decode() {
```
Comment (battledash) [inline maxima-ui/src/fs/image_loader.rs]: You already know what I'm gonna say PepeLaugh

File: maxima-ui/src/ui_image.rs [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+    pub size: Vec2,
+}
+
+#[derive(Clone, PartialEq)]
+pub enum GameImageType {
+    Hero,
+    Logo,
+}
+
+pub async fn write_image_at_url_to_file(url: String, file_name: &PathBuf) -> Result<()>{
```
Comment (battledash) [inline maxima-ui/src/ui_image.rs]: `download_image`

File: maxima-ui/src/ui_image.rs [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+    info!("Downloading image at {:?}", &url);
+    let result = reqwest::get(&url).await;
+
```
Comment (battledash) [inline maxima-ui/src/ui_image.rs]: Remove this newline

File: maxima-ui/src/ui_image.rs [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+    if result.is_err() {
+        bail!("Failed to download {}! Reason: {:?}", &url, &result);
+    }
+    let body= result?.bytes().await?;
```
Comment (battledash) [inline maxima-ui/src/ui_image.rs]: Add a newline before this and a space after body

File: maxima-ui/src/ui_image.rs [hunk 30]
```diff
@@ -0,0 +1,97 @@
...
+    let file = File::create(&file_name).await;
+
```
Comment (battledash) [inline maxima-ui/src/ui_image.rs:37]: Remove this newline too

File: maxima-ui/src/ui_image.rs [hunk 31]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    let copy_result = io::copy(&mut body.as_slice(), &mut file?).await;
+    if copy_result.is_ok() {
+        debug!("Copied file!")
+    } else {
+        error!("Failed to copy file! Reason: {:?}", copy_result.err())
+    }
+
+    Ok(())
```
Comment (battledash) [inline maxima-ui/src/ui_image.rs]:
```rust
    if let Err(err) = io::copy(&mut body.as_slice(), &mut file?).await {
        error!("Failed to copy file! Reason: {:?}", err)
    }

    debug!("Copied file!")
    Ok(())
```

**Review by battledash — CHANGES_REQUESTED** [review]

File: maxima-ui/src/bridge/game_images.rs (hunk 15 above)
Comment (headassbtw) [reply maxima-ui/src/bridge/game_images.rs]: LMAO

**Review by battledash — APPROVED** [review]

**Pull request merged**